### PR TITLE
Removed auto-constructors for finicky objects

### LINKED ISSUE #131: Hide object inits from the objects themselves and enforce creation via Create methods
#### Description:

Validation should be applied to the objects. This can be done by adding the `[NoAutoConstructor]` decorator to the object to be hidden (at least within a BHoM-UI) and having methods like this:

```c#
namespace BH.Engine.LadybugTools
{
    public static partial class Create
    {
        [Description("Create a Shelter object.")]
        [Input("vertices", "The vertices of the shelter.")]
        [Input("windPorosity", "The hourly wind porosity of the shelter (0-1).")]
        [Input("radiationPorosity", "The hourly radiation porosity of the shelter (0-1).")]
        [Output("shelter", "A Shelter object.")]
        public static Shelter Shelter(List<Point> vertices, List<double> windPorosity = null, List<double> radiationPorosity = null)
        {
            if ((windPorosity.Count() == 0 && windPorosity.Sum() == 0) || windPorosity == null)
                windPorosity = Enumerable.Repeat(0.0, 8760).ToList();

            if (windPorosity.Count != 8760)
            {
                BH.Engine.Base.Compute.RecordError("windPorosity must be a list of 8760 values");
                return null;
            }
            if (windPorosity.Where(x => x < 0 || x > 1).Any())
            {
                BH.Engine.Base.Compute.RecordError("Shelter wind porosity must be between 0 and 1.");
                return null;
            }

            if ((radiationPorosity.Count() == 0 && radiationPorosity.Sum() == 0) || radiationPorosity == null)
                radiationPorosity = Enumerable.Repeat(0.0, 8760).ToList();

            if (radiationPorosity.Count != 8760)
            {
                BH.Engine.Base.Compute.RecordError("radiationPorosity must be a list of 8760 values");
                return null;
            }

            if (radiationPorosity.Where(x => x < 0 || x > 1).Any())
            {
                BH.Engine.Base.Compute.RecordError("Shelter radiation porosity must be between 0 and 1.");
                return null;
            }

            if (radiationPorosity.Sum() + windPorosity.Sum() == 8760 * 2)
            {
                BH.Engine.Base.Compute.RecordError("This shelter is completely transmissive and would have no effect.");
                return null;
            }

            if (!BH.Engine.Geometry.Create.Polyline(vertices).IsPlanar())
            {
                BH.Engine.Base.Compute.RecordError("Shelter vertices are not planar.");
                return null;
            }

            return new Shelter()
            {
                Vertices = vertices,
                WindPorosity = windPorosity,
                RadiationPorosity = radiationPorosity
            };
        }
    }
}

These exist on a throwaway branch currently, so if you fancy coding this up get in touch with me and I can share it with you to save on having to write it again.
```

### PR DESCRIPTION
**Pull request by tg359**
### Issues addressed by this PR

Closes #131 

`[NoAutoConstructor]` added to objects where they are either created via a process, or have some validation required in order to be suitable for use in downstream processes.

### CODE REVIEW

File: LadybugTools_Engine/Create/AnalysisPeriod.cs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        [Input("stMonth", "The starting month of the analysis period. 1-12")]
+        [Input("stDay", "The starting day of the analysis period. 1-31")]
+        [Input("stHour", "The starting hour of the analysis period. 0-23")]
+        [Input("endMonth", "The ending month of the analysis period. 1-12")]
+        [Input("endDay", "The ending day of the analysis period. 1-31")]
+        [Input("endHour", "The ending hour of the analysis period. 0-23")]
+        [Input("isLeapYear", "Whether the analysis period represents a leap year.")]
+        [Input("timestep", "The number of timesteps per hour. One of {1, 2, 3, 4, 5, 6, 10, 12, 15, 20, 30, 60}")]
+        [Output("analysisPeriod", "An AnalysisPeriod object.")]
+        public static AnalysisPeriod AnalysisPeriod(int stMonth = 1, int stDay = 1, int stHour = 0, int endMonth = 12, int endDay = 31, int endHour = 23, bool isLeapYear = false, int timestep = 1)
```
Comment (Tom-Kingstone) [inline LadybugTools_Engine/Create/AnalysisPeriod.cs:42]: seeing as these variable names aren't used by the serialiser, could they be clearer? (eg. `stMonth` to `startMonth`)

File: LadybugTools_Engine/Create/EnergyMaterial.cs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        [Input("thickness", "The thickness of the material in meters.")]
+        [Input("conductivity", "The conductivity of the material in W/m-K.")]
+        [Input("density", "The density of the material in kg/m3.")]
+        [Input("specificHeat", "The specific heat of the material in J/kg-K.")]
+        [Input("roughness", "The roughness of the material.")]
+        [Input("thermalAbsorptance", "The thermal absorptance of the material.")]
+        [Input("solarAbsorptance", "The solar absorptance of the material.")]
+        [Input("visibleAbsorptance", "The visible absorptance of the material.")]
+        [Output("energyMaterial", "An EnergyMaterial object.")]
+        public static EnergyMaterial EnergyMaterial(
```
Comment (Tom-Kingstone) [inline LadybugTools_Engine/Create/EnergyMaterial.cs:42]: Please include units in the input message strings.

File: LadybugTools_Engine/Create/EnergyMaterialVegetation.cs [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        [Input("roughness", "The roughness of the material.")]
+        [Input("plantHeight", "The height of the vegetation in meters.")]
+        [Input("leafAreaIndex", "The leaf area index of the vegetation.")]
+        [Input("leafReflectivity", "The reflectivity of the leaves.")]
+        [Input("leafEmissivity", "The emissivity of the leaves.")]
+        [Input("minStomatalResist", "The minimum stomatal resistance of the leaves.")]
+        [Input("soilThermalAbsorptance", "The thermal absorptance of the soil.")]
+        [Input("soilSolarAbsorptance", "The solar absorptance of the soil.")]
+        [Input("soilVisibleAbsorptance", "The visible absorptance of the soil.")]
+        [Output("energyMaterialVegetation", "An EnergyMaterialVegetation object.")]
```
Comment (Tom-Kingstone) [inline LadybugTools_Engine/Create/EnergyMaterialVegetation.cs:46]: Units again

File: LadybugTools_Engine/Create/Location.cs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    public static partial class Create
+    {
+        [Description("Create a Location object.")]
+        [Input("city", "The city of the location.")]
+        [Input("state", "The state of the location.")]
+        [Input("country", "The country of the location.")]
+        [Input("latitude", "The latitude of the location.")]
+        [Input("longitude", "The longitude of the location.")]
+        [Input("timeZone", "The time zone of the location.")]
+        [Input("elevation", "The elevation of the location.")]
```
Comment (Tom-Kingstone) [inline LadybugTools_Engine/Create/Location.cs:39]: same here

File: LadybugTools_Engine/Create/Typology.cs [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+namespace BH.Engine.LadybugTools
+{
+    public static partial class Create
+    {
+        [Description("Create a Typology object.")]
+        [Input("identifier", "The identifier of the typology.")]
+        [Input("shelters", "The shelters of the typology.")]
+        [Input("evaporativeCoolingEffect", "The hourly evaporative cooling effect of the typology (0-1).")]
+        [Input("targetWindSpeed", "The hourly target wind speed of the typology.")]
+        [Input("radiantTemperatureAdjustment", "The hourly radiant temperature adjustment of the typology.")]
```
Comment (Tom-Kingstone) [inline LadybugTools_Engine/Create/Typology.cs:38]: Units again

**Review by Tom-Kingstone — CHANGES_REQUESTED** [review]
Some minor changes needed

File: LadybugTools_Engine/Create/AnalysisPeriod.cs (hunk 1 above)
Comment (FraserGreenroyd) [reply LadybugTools_Engine/Create/AnalysisPeriod.cs:42]: Seconded 👍

File: LadybugTools_Engine/Create/EnergyMaterial.cs (hunk 2 above)
Comment (FraserGreenroyd) [reply LadybugTools_Engine/Create/EnergyMaterial.cs:42]: Providing we have the quantities available we should use them but if we don't have them in Quantities oM then we could ignore for now.

File: LadybugTools_Engine/Create/EnergyMaterial.cs [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+            double specificHeat,
+            Roughness roughness = Roughness.MediumRough,
+            double thermalAbsorptance = 0.9,
+            double solarAbsorptance = 0.7,
+            double visibleAbsorptance = 0.7
+        )
+        {
+            if (identifier.Contains("!"))
+            {
+                BH.Engine.Base.Compute.RecordError("identifier cannot contain '!' character");
```
Comment (FraserGreenroyd) [inline LadybugTools_Engine/Create/EnergyMaterial.cs:56]:
```suggestion
                BH.Engine.Base.Compute.RecordError("Identifier cannot contain '!' character");
```

File: LadybugTools_Engine/Create/EnergyMaterial.cs [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+                return null;
+            }
+
+            if (identifier.Length > 100)
+            {
+                BH.Engine.Base.Compute.RecordError("identifier cannot be longer than 100 characters");
```
Comment (FraserGreenroyd) [inline LadybugTools_Engine/Create/EnergyMaterial.cs:62]:
```suggestion
                BH.Engine.Base.Compute.RecordError("Identifier cannot be longer than 100 characters");
```

File: LadybugTools_Engine/Create/EnergyMaterial.cs [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+                return null;
+            }
+
+            if (specificHeat < 100)
+            {
+                BH.Engine.Base.Compute.RecordError("specificHeat must be greater than 100");
```
Comment (FraserGreenroyd) [inline LadybugTools_Engine/Create/EnergyMaterial.cs:68]:
```suggestion
                BH.Engine.Base.Compute.RecordError("SpecificHeat must be greater than 100");
```

File: LadybugTools_Engine/Create/EnergyMaterial.cs [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+                return null;
+            }
+
+            if (conductivity < 0)
+            {
+                BH.Engine.Base.Compute.RecordError("conductivity must be greater than 0");
```
Comment (FraserGreenroyd) [inline LadybugTools_Engine/Create/EnergyMaterial.cs:74]:
```suggestion
                BH.Engine.Base.Compute.RecordError("Conductivity must be greater than 0");
```

File: LadybugTools_Engine/Create/EnergyMaterial.cs [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+                return null;
+            }
+
+            if (density < 0)
+            {
+                BH.Engine.Base.Compute.RecordError("density must be greater than 0");
```
Comment (FraserGreenroyd) [inline LadybugTools_Engine/Create/EnergyMaterial.cs:80]:
```suggestion
                BH.Engine.Base.Compute.RecordError("Density must be greater than 0");
```

File: LadybugTools_Engine/Create/EnergyMaterial.cs [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+                return null;
+            }
+
+            if (thickness <= 0)
+            {
+                BH.Engine.Base.Compute.RecordError("thickness must be greater than 0");
```
Comment (FraserGreenroyd) [inline LadybugTools_Engine/Create/EnergyMaterial.cs:86]:
```suggestion
                BH.Engine.Base.Compute.RecordError("Thickness must be greater than 0");
```

File: LadybugTools_Engine/Create/EnergyMaterial.cs [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+                return null;
+            }
+
+            if (roughness == Roughness.Undefined)
+            {
+                BH.Engine.Base.Compute.RecordError("roughness must be defined");
```
Comment (FraserGreenroyd) [inline LadybugTools_Engine/Create/EnergyMaterial.cs:92]:
```suggestion
                BH.Engine.Base.Compute.RecordError("Roughness must be defined");
```

File: LadybugTools_Engine/Create/EnergyMaterial.cs [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+                return null;
+            }
+
+            if (thermalAbsorptance < 0 || thermalAbsorptance > 1 || solarAbsorptance < 0 || solarAbsorptance > 1 || visibleAbsorptance < 0 || visibleAbsorptance > 1)
+            {
+                BH.Engine.Base.Compute.RecordError("thermalAbsorptance, solarAbsorptance, and visibleAbsorptance must be between 0 and 1");
```
Comment (FraserGreenroyd) [inline LadybugTools_Engine/Create/EnergyMaterial.cs:98]:
```suggestion
                BH.Engine.Base.Compute.RecordError("ThermalAbsorptance, SolarAbsorptance, and VisibleAbsorptance must be between 0 and 1");
```

File: LadybugTools_Engine/Create/EnergyMaterialVegetation.cs [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+            double leafEmissivity = 0.95,
+            double minStomatalResist = 180,
+            double soilThermalAbsorptance = 0.9,
+            double soilSolarAbsorptance = 0.7,
+            double soilVisibleAbsorptance = 0.7
+        )
+        {
+            if (identifier.Contains("!"))
+            {
+                BH.Engine.Base.Compute.RecordError("identifier cannot contain '!' character");
```
Comment (FraserGreenroyd) [inline LadybugTools_Engine/Create/EnergyMaterialVegetation.cs:66]: Please capitalise first letter of error messages 😄

File: LadybugTools_Engine/Create/Location.cs [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+                return null;
+            }
+
+            if (timeZone < -12 || timeZone > 14)
+            {
+                BH.Engine.Base.Compute.RecordError("timeZone must be within -12 to 14");
+                return null;
+            }
+
+
```
Comment (FraserGreenroyd) [inline LadybugTools_Engine/Create/Location.cs:79]:
```suggestion
```

File: LadybugTools_oM/EnergyMaterial.cs [hunk 16]
```diff
@@ -22,39 +22,41 @@
...
+
         [Description("The name of this EnergyMaterial.")]
-        public virtual string Identifier { get; set; } = "Generic Opaque Material";
-        
+        public virtual string Identifier { get; set; }
+
         [Description("Thickness of material (m).")]
-        public virtual double Thickness { get; set; } = 0.1;
-        
+        public virtual double Thickness { get; set; }
```
Comment (FraserGreenroyd) [inline LadybugTools_oM/EnergyMaterial.cs:39]:
```suggestion
        [Length]
        public virtual double Thickness { get; set; }
```

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]

File: LadybugTools_Engine/Create/AnalysisPeriod.cs (hunk 1 above)
Comment (tg359) [reply LadybugTools_Engine/Create/AnalysisPeriod.cs:42]: But they are used during serialisation as the Python-side is expecting an AnalysisPeriod dict with key names per those given on the objects currently. If we hide the functionality within an adapter then I'm happy to change these on BHoM-side and have it rename upon conversion to a JSON-string ... but for now it's as simple as `obj.ToJson()` without any additional code required.

Comment (Tom-Kingstone) [reply LadybugTools_Engine/Create/AnalysisPeriod.cs:43]: The `AnalysisPeriod` object itself can stay the same, as the object is what is used for serialisation. The serialiser doesn't see this method - rather the object itself - so variable names in this method can be anything as long as the object stays the same.

Comment (tg359) [reply LadybugTools_Engine/Create/AnalysisPeriod.cs:43]: > The `AnalysisPeriod` object itself can stay the same, as the object is what is used for serialisation. The serialiser doesn't see this method - rather the object itself - so variable names in this method can be anything as long as the object stays the same.

Good point! I'll make that change to this method.

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]
Project compliance will fail so let's get the bot to fix that after `Typology`.

@BHoMBot check project-compliance

**Comment by Tom-Kingstone** [comment]
@BHoMBot check project-compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot fix project file ref. 18659371395

**Comment by FraserGreenroyd** [comment]
@BHoMBot check compliance

**Comment by Tom-Kingstone** [comment]
@BHoMBot check compliance
@BHoMBot check required

**Review by FraserGreenroyd — APPROVED** [review]
All good on code review with comments resolved, thanks @Tom-Kingstone for picking up the additional `NoAutoConstructor` discussed with @tg359 on Teams.

Happy to deploy for further testing.

**Review by Tom-Kingstone — DISMISSED** [review]
approved as @FraserGreenroyd will set out

**Comment by FraserGreenroyd** [comment]
@BHoMBot check compliance
@BHoMBot check required

**Review by FraserGreenroyd — APPROVED** [review]
Reapproving based on [this comment](https://github.com/BHoM/LadybugTools_Toolkit/pull/135#pullrequestreview-1729447079) following a rebase to resolve conflicts.

**Comment by FraserGreenroyd** [comment]
@BHoMBot this is a DevOps instruction. I am requesting neutral checks on: installer, unit-tests, ready-to-merge

**Pull request merged**